azure/scope/machine.go

```diff
@@ -145,13 +145,27 @@ func (m *MachineScope) InboundNatSpecs() []azure.InboundNatSpec {
 }
 
 // NICSpecs returns the network interface specs.
-func (m *MachineScope) NICSpecs() []azure.NICSpec {
+func (m *MachineScope) NICSpecs() ([]azure.NICSpec, error) {
+	subnetName := m.AzureMachine.Spec.SubnetName
+	if subnetName == "" {
+		subnets := m.Subnets()
+		var subnetCount int
+		for _, subnet := range subnets {
+			if string(subnet.Role) == m.Role() {
+				subnetCount++
+				subnetName = subnet.Name
+			}
+		}
+		if subnetCount == 0 || subnetCount > 1 {
+			return []azure.NICSpec{}, errors.New("a subnet name must be specified when no subnets are specified or more than 1 subnet of the same role exist")
+		}
+	}
```

Comment:
This code is required while the new `SubnetName` field is optional. With this code we try to calculate a sane default for the subnet name. We fail to do that if there are no subnets (this should not happen) or the user has specified more than 1 subnet with the `node` role.

Reply:
see my comment above, we could add validation in webhooks to avoid this situation and then always set subnetName to the first (and only) node subnet if it's empty for back compat

Reply:
I don't see how to access the `Subnets` from the `AzureMachine` validation. There seems we can only access the `AzureMachine` object when validating, so I don't see how to check the number of subnets defined in the `AzureCluster` object. Any pointers?

Reply:
Ok so adding webhook validation might be difficult/undesirable because it's across 2 objects as you mentioned so leaving it in scope is fine but would you consider extracting it to it's own func and actually defaulting the subnet name at the beginning of reconciliation if it's not set (similar to what we do for `SetDNSName` in cluster scope)

Reply:
Done, PTAL.